### ISSUE TITLE
Mnemosyne just thrashed my drives

### ISSUE THREAD

**Issue by indiv0** [mnemosyne-proj/mnemosyne#99]
# Description

~3 days ago I wanted to give mnemosyne a try, so I installed the package from the Arch Linux AUR.
I changed one or two settings, but then I had to leave, so I left mnemosyne open without having created any cards.

Today, I noticed that my PC was responding slowly, so I decided to investigate.
According to `iotop`, `mnemosyne` was still open and was writing to my NVMe (root) drive at ~300 MB/s.
I immediately killed the process. Unfortunately, I didn't think to take a screenshot of `iotop` at the time, or run `strace`.

After doing some investigation (see Debug Info, below), it turned out that since the last reboot, ~49.73 TB had been written to my NVMe drive and ~550 GB had been written to one of my HDD drives. Since my PC has been up for ~14 days, I cannot attribute all of this to mnemosyne (I can attribute at least ~150 GB of the data written to the HDD to other programs). However, it still appears that **mnemosyne wrote tens of terabytes to my near-unused NVMe drive**, using up ~1/4 of its rated write endurance in the process. For reference, prior to this, I had written <5 TB total to the drive over the past *year*.

Attempting to open mnemosyne now gives the following error:

```
An unexpected error has occurred.
Please forward the following info to the developers:

Traceback (innermost last):
  File "/usr/bin/mnemosyne", line 262, in <module>
    debug_file=options.debug_file)
  File "/usr/lib/python3.7/site-packages/mnemosyne/libmnemosyne/__init__.py", line 193, in initialise                                               
    self.activate_components()
  File "/usr/lib/python3.7/site-packages/mnemosyne/libmnemosyne/__init__.py", line 282, in activate_components                                      
    self.component_manager.current(component).activate()
  File "/usr/lib/python3.7/site-packages/mnemosyne/libmnemosyne/logger.py", line 23, in activate                                                    
    self.archive_old_log()
  File "/usr/lib/python3.7/site-packages/mnemosyne/libmnemosyne/logger.py", line 192, in archive_old_log                                            
    f.write(l.encode("utf-8"))
  File "/usr/lib/python3.7/bz2.py", line 241, in write
    compressed = self._compressor.compress(data)
 KeyboardInterrupt
```

I've decided to open an issue to attempt to resolve this so no one else experiences this.
If any further debug information is required, let me know and I will provide it at the next opportunity.
Quite frankly, I'm furious right now and not in the mood to debug this further at the current moment.

# Debug Info

* `uname -a`
```
Linux sputnik 4.18.16-arch1-1-ARCH #1 SMP PREEMPT Sat Oct 20 22:06:45 UTC 2018 x86_64 GNU/Linux
```

* `pacman -Q | grep mnemosyne`
```
mnemosyne 2.6.1-1
```

* `uptime`
```
 02:08:39 up 14 days, 27 min, 16 users,  load average: 1.73, 2.11, 2.37
```

* `df -T`

```
Filesystem                Type     1K-blocks      Used Available Use% Mounted on
dev                       devtmpfs   5804948         0   5804948   0% /dev
run                       tmpfs     16336184      1012  16335172   1% /run
/dev/mapper/vgroup-lvroot ext4     234147832  38469808 183714224  18% /
tmpfs                     tmpfs     16336184    262628  16073556   2% /dev/shm
tmpfs                     tmpfs     16336184         0  16336184   0% /sys/fs/cgroup
tmpfs                     tmpfs     16336184    407456  15928728   3% /tmp
/dev/nvme0n1p1            vfat       1046512     65692    980820   7% /boot
tank01                    zfs       25908352         0  25908352   0% /mnt/tank01
tank01/vm                 zfs       25908452        96  25908356   1% /mnt/tank01/vm
tank01/vm/windows         zfs       74854680  48946324  25908356  66% /mnt/tank01/vm/windows
tank02                    zfs      603361536 112892672 490468864  19% /mnt/tank02
tank02/tmp                zfs      665136128 174667776 490468352  27% /mnt/tank02/tmp
tmpfs                     tmpfs      3267236        32   3267204   1% /run/user/1000
```

* `awk '/nv/ {print $3"\t"$10 / 2 / 1024}' /proc/diskstats`

```
nvme0n1 4.97347e+07
nvme0n1p1       56.9048
nvme0n1p2       4.97347e+07
```

* `awk '/sd/ {print $3"\t"$10 / 2 / 1024}' /proc/diskstats`

```
sda     550965
sda1    550965
sdb     1659.93
sdb1    1659.93
```

* `sudo smartctl -A /dev/nvme0n1`

```
...
=== START OF SMART DATA SECTION ===
SMART/Health Information (NVMe Log 0x02, NSID 0x1)
Critical Warning:                   0x00
Temperature:                        38 Celsius
Available Spare:                    100%
Available Spare Threshold:          10%
Percentage Used:                    18%
Data Units Read:                    3,204,202 [1.64 TB]
Data Units Written:                 107,146,899 [54.8 TB]
Host Read Commands:                 47,591,909
Host Write Commands:                176,159,535
Controller Busy Time:               2,202
Power Cycles:                       450
Power On Hours:                     738
Unsafe Shutdowns:                   32
Media and Data Integrity Errors:    0
Error Information Log Entries:      26
Warning  Comp. Temperature Time:    0
Critical Comp. Temperature Time:    0
Temperature Sensor 1:               38 Celsius
Temperature Sensor 2:               44 Celsius
```

**Comment by pbienst** [comment]
I'm sorry to hear that your furious. 

However, I've never seen anything like this, and I cannot really think of any way Mnemosyne would generate Tb's of data from an empty database. Which were the settings that you changed? What are the filenames that were written to? Mnemosyne should only ever write in ~/.mnemosyne and ~/.config/mnemosyne.

So, in order to debug this, I would need more information. Please delete the two directories above, start the program again and monitor for any strange behaviour. It would also be best to try and install Mnemosyne directly from the source tarball we provide, as sometimes packagers for distributions make modifications to the code we don't have any control over.

**Comment by indiv0** [comment]
Unfortunately I can't remember what settings I changed. Is there a way to diff the existing settings file vs. the default? I'm not exactly sure how to determine what files were written to either, but I will examine those directories for a start.

The most suspect thing I can see at the moment is that `./local/share/mnemosyne/default.db` is 314 MB, which is curious.

Additionally, `.local/share/mnemosyne/backups` has 10 backup files, all 315 MB, 8 of which have the same modification time (down to the minute). I believe that 1:18 AM is approximately when I terminated the process, as well. This leads to me to believe that there was an error in the backup logic.

* `ls -lah .local/share/mnemosyne/backups/`

```
total 3.1G
drwxr-xr-x 2 indiv0 indiv0 4.0K Nov 12 03:28 ./
drwxr-xr-x 6 indiv0 indiv0 4.0K Nov 12 02:19 ../
-rw-r--r-- 1 indiv0 indiv0 314M Nov 12 01:18 default-20181112-011810.db
-rw-r--r-- 1 indiv0 indiv0 314M Nov 12 01:18 default-20181112-011811.db
-rw-r--r-- 1 indiv0 indiv0 314M Nov 12 01:18 default-20181112-011812.db
-rw-r--r-- 1 indiv0 indiv0 314M Nov 12 01:18 default-20181112-011814.db
-rw-r--r-- 1 indiv0 indiv0 314M Nov 12 01:18 default-20181112-011816.db
-rw-r--r-- 1 indiv0 indiv0 314M Nov 12 01:18 default-20181112-011817.db
-rw-r--r-- 1 indiv0 indiv0 314M Nov 12 01:18 default-20181112-011818.db
-rw-r--r-- 1 indiv0 indiv0 314M Nov 12 01:18 default-20181112-011819.db
-rw-r--r-- 1 indiv0 indiv0 314M Nov 12 02:19 default-20181112-021952.db
-rw-r--r-- 1 indiv0 indiv0 314M Nov 12 03:28 default-20181112-032834.db
```

As for the installation itself, the AUR package was out-of-date, so I used the following updated [PKGBUILD](https://gist.github.com/f0ea7000b121fe6bad6c6b50f18256ae) file to install it. However, as can be seen from a cursory examination, that PKGBUILD performs no out of the ordinary modifications, and installs mnemosyne from the [2.6.1 source tarball](http://downloads.sourceforge.net/mnemosyne-proj/Mnemosyne-2.6.1.tar.gz), and verified by its sha256sum output: `6331241bc7caf26fed2e4d32c3c18da33116389b67af621da679a950f08c9c75`.

I will backup the directories then delete them and monitor for suspicious behaviour tomorrow, to see if this occurs again.

**Comment by pbienst** [comment]
Thanks for the swift feedback!

The settings are written in config.db and config.py. If you backup those, they can be compared against the default values that will get generated if you start from an empty directory.

Main one to look out for is day_starts_at in config.py, as at that point in time, a backup gets created. (These settings are not accessible from the GUI though, as only power users edit this file.)

Also, there was no daylight saving time switch in your timezone when this happened? Or other weird things happening to your clock? Or an impossibility for Mnemosyne to write the time of the last backup to config.db?

**Comment by indiv0** [comment]
* [config.py](https://gist.github.com/indiv0/2729e80b41180bae99ec1c89ba3e8acc)
* [config.db](https://drive.google.com/file/d/1Y-G0msI7TTx1h2wDWhlyhed9YUHt1fCX/view?usp=sharing)

Nothing appears out of the ordinary to me in either. I'll see if I can perform a full diff of `config.db` tomorrow.
In `config.py`, `day_starts_at` is set to `3`.

Per my `/var/log/pacman.log`, I installed mnemosyne at `2018-11-07 19:54`, EST. Meaning that I was actually incorrect about when I'd installed mnemosyne and how long it had been running for (i.e. 5 days instead of 3). In my area, the DST switch occurs at `2018-11-04 02:00`, EST. So the switch occurred 3 days prior to installation, but the computer has been on since before the switch occurred. The clock is currently set correctly, and I have not manually changed it. The machine is not running an `ntp` daemon. I can't think of anything else that might have affected the clock. The drive itself is only at ~20% of capacity, so it's unlikely that there was a space issue. Permissions all look correct as well. The only other thing I can think of is that I have had some internet connectivity issues over this time period.

**Comment by pbienst** [comment]
Right, then the only thing to do is to see if this is reproducible from a empty Mnemosyne data directory...

**Comment by indiv0** [comment]
Alright I'll try that.

For reference, I also examined my system journal to see if anything obvious popped out. My 2 runs of mnemosyne after I noticed the issue gave the following errors:

```
Nov 12 02:19:17 sputnik python[20409]: libpng warning: iCCP: known incorrect sRGB profile
Nov 12 02:19:17 sputnik python[20409]: libpng warning: iCCP: known incorrect sRGB profile
Nov 12 02:19:17 sputnik python[20409]: libpng warning: iCCP: known incorrect sRGB profile
Nov 12 02:19:17 sputnik python[20409]: libpng warning: iCCP: known incorrect sRGB profile
Nov 12 02:19:17 sputnik python[20409]: libpng warning: iCCP: known incorrect sRGB profile
Nov 12 02:19:39 sputnik python[20409]: QIODevice::write (QProcess): device not open
Nov 12 02:19:39 sputnik python[20409]: QIODevice::write (QProcess): device not open
Nov 12 02:19:52 sputnik python[20409]: QIODevice::write (QProcess): device not open
Nov 12 02:19:52 sputnik python[20409]: QIODevice::write (QProcess): device not open
Nov 12 03:27:37 sputnik python[14851]: libpng warning: iCCP: known incorrect sRGB profile
Nov 12 03:27:37 sputnik python[14851]: libpng warning: iCCP: known incorrect sRGB profile
Nov 12 03:27:37 sputnik python[14851]: libpng warning: iCCP: known incorrect sRGB profile
Nov 12 03:27:37 sputnik python[14851]: libpng warning: iCCP: known incorrect sRGB profile
Nov 12 03:27:37 sputnik python[14851]: libpng warning: iCCP: known incorrect sRGB profile
Nov 12 03:27:38 sputnik python[14851]: QIODevice::write (QProcess): device not open
Nov 12 03:27:38 sputnik python[14851]: QIODevice::write (QProcess): device not open
...
```

But there don't appear to be any python errors from before then.

`systemd-timesyncd` actually failed to startup:

```

Nov 07 19:59:55 sputnik dbus-daemon[2523]: [session uid=1000 pid=2523] Reloaded configuration
Nov 07 19:59:55 sputnik dbus-daemon[1738]: [system] Reloaded configuration
Nov 07 19:59:55 sputnik dbus-daemon[1738]: [system] Reloaded configuration
Nov 07 19:59:55 sputnik dbus-daemon[1738]: [system] Reloaded configuration
Nov 07 19:59:56 sputnik dbus-daemon[1738]: [system] Reloaded configuration
Nov 07 19:59:56 sputnik dbus-daemon[1738]: Encountered error 'Error in file /usr/share/dbus-1/system.d/org.freedesktop.systemd1.conf, line 293, column 16: unclosed token
                                           ' while parsing '/usr/share/dbus-1/system.d/org.freedesktop.systemd1.conf'
Nov 07 19:59:56 sputnik dbus-daemon[1738]: dbus-daemon[1738]: Encountered error 'Error in file /usr/share/dbus-1/system.d/org.freedesktop.systemd1.conf, line 293, column 16: unclosed token
Nov 07 19:59:56 sputnik dbus-daemon[1738]: ' while parsing '/usr/share/dbus-1/system.d/org.freedesktop.systemd1.conf'
Nov 07 19:59:56 sputnik dbus-daemon[1738]: [system] Reloaded configuration
Nov 07 19:59:56 sputnik dbus-daemon[1738]: dbus-daemon[1738]: [system] Reloaded configuration
Nov 07 19:59:56 sputnik dbus-daemon[1738]: Unknown username "systemd-timesync" in message bus configuration file
Nov 07 19:59:56 sputnik dbus-daemon[1738]: dbus-daemon[1738]: Unknown username "systemd-timesync" in message bus configuration file
Nov 07 19:59:56 sputnik dbus-daemon[1738]: [system] Reloaded configuration
Nov 07 19:59:56 sputnik dbus-daemon[1738]: dbus-daemon[1738]: [system] Reloaded configuration
...
Nov 07 19:59:56 sputnik systemd-sysusers[8211]: Creating group systemd-timesync with gid 970.
Nov 07 19:59:56 sputnik systemd-sysusers[8211]: Creating user systemd-timesync (systemd Time Synchronization) with uid 970 and gid 970.
Nov 07 19:59:56 sputnik systemd[1]: Reexecuting.
Nov 07 19:59:56 sputnik kernel: systemd: 39 output lines suppressed due to ratelimiting
Nov 07 19:59:56 sputnik systemd[1]: systemd 239 running in system mode. (+PAM +AUDIT -SELINUX -IMA -APPARMOR +SMACK -SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=hybrid)
Nov 07 19:59:56 sputnik systemd[1]: Detected architecture x86-64.
Nov 07 20:00:12 sputnik dbus-daemon[2523]: [session uid=1000 pid=2523] Reloaded configuration
Nov 07 20:00:12 sputnik dbus-daemon[2523]: [session uid=1000 pid=2523] Reloaded configuration
Nov 07 20:00:12 sputnik dbus-daemon[2523]: [session uid=1000 pid=2523] Reloaded configuration
Nov 07 20:00:12 sputnik dbus-daemon[2523]: [session uid=1000 pid=2523] Reloaded configuration
Nov 07 20:00:12 sputnik dbus-daemon[2523]: [session uid=1000 pid=2523] Reloaded configuration
Nov 07 20:00:12 sputnik dbus-daemon[2523]: [session uid=1000 pid=2523] Reloaded configuration
Nov 07 20:00:31 sputnik systemd[1]: Reloading.
```

**Comment by pbienst** [comment]
A clock drift by itself should be harmless.

The lines in the log you see referring to 'database loaded' in the future are a hack so that the statistics know about the number of cards scheduled in the future on a particular day.

**Comment by pbienst** [comment]
Just wondering if you've managed to reproduce this?

**Comment by indiv0** [comment]
Sorry for the late reply, I've been busy lately. I tried to reproduce this several times, but was unable to. I think the symptoms I observed could be explained by the condition on [this](https://github.com/mnemosyne-proj/mnemosyne/blob/master/mnemosyne/libmnemosyne/controllers/default_controller.py#L49) line perpetually evaluating to `True` (in fact, setting this to `True` and running mnemosyne results in the observed behaviour), but I can't figure out how that might come about in normal operation.

Some testing of the [`start_of_day_n_days_ago(self, n)`](https://github.com/mnemosyne-proj/mnemosyne/blob/c2d0aa4a69021907188863a65ff96bed5874a5d2/mnemosyne/libmnemosyne/databases/SQLite_statistics.py#L150-L158) (which is used to set `self.next_rollover`) reveals that it appears to work incorrectly for timestamps which fall in the hour immediately following (IIRC) the DST switchover, and the conditional statement will evaluate to `True` for every second of that hour (resulting in 3600 consecutive backups being made). However while this explanation is near perfect, it doesn't account for the fact that backups continued to occur past that hour, and that I noticed this several days after the DST switchover should have happened (though perhaps my computer did not do it at the correct time for my locale?).

My next idea is to try setting the hardware clock to around that date manually and letting mnemosyne run to try and reproduce the issue.
